akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -107,6 +107,8 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
           //assuming there can be no further in messages
           downstreamStatus.set(Canceled)
           dataQueue.clear()
+          // if blocked reading, make sure the take() completes
+          dataQueue.put(ByteString())
```

Comment:
Or do you mean the unblocking? The future below is blockingly waiting on a `take()`, so we send an element through the queue to make it unblock and notice that the stage is done with its business.